beacon_node/eth2_libp2p/src/rpc/codec/ssz_snappy.rs

```diff
@@ -409,11 +412,50 @@ impl<TSpec: EthSpec> OutboundCodec<RPCRequest<TSpec>> for SSZSnappyOutboundCodec
                 )?)))
             }
             Err(e) => match e.kind() {
-                // Haven't received enough bytes to decode yet
-                // TODO: check if this is the only Error variant where we return `Ok(None)`
+                // Haven't received enough bytes to decode yet, wait for more
                 ErrorKind::UnexpectedEof => Ok(None),
                 _ => Err(e).map_err(RPCError::from),
             },
         }
     }
 }
+
+/// Wrapper over `read` implementation of `FrameDecoder`.
+///
+/// Works like the standard `read_exact` implementation, except that it returns an error if length of bytes
+/// read from the underlying reader is greater than `max_compressed_length`.  
+fn read_exact<T: std::convert::AsRef<[u8]>>(
+    reader: &mut FrameDecoder<Cursor<T>>,
+    mut buf: &mut [u8],
+    max_compressed_len: u64,
+) -> Result<(), std::io::Error> {
+    let mut i = reader.get_ref().position();
+    let mut count = 0;
+    while !buf.is_empty() {
+        match reader.read(buf) {
+            Ok(0) => break,
+            Ok(n) => {
+                let tmp = buf;
+                buf = &mut tmp[n..];
+            }
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
+            Err(e) => return Err(e),
+        }
+        count += reader.get_ref().position() - i;
```

Comment:
can the position here ever be less than the previous position and underflow?

Reply:
Checked the snappy read implementation and it always advances the reader forward as far as I can tell. However, added an underflow check to be safe.